airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.featureflag
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import com.launchdarkly.sdk.ContextKind
+import com.launchdarkly.sdk.LDContext
+import com.launchdarkly.sdk.LDUser
+import com.launchdarkly.sdk.server.LDClient
+import java.lang.Thread.MIN_PRIORITY
+import java.nio.file.Path
+import java.nio.file.StandardWatchEventKinds
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.thread
+import kotlin.concurrent.write
+import kotlin.io.path.isRegularFile
+
+/**
+ * Open Questions:
+ * 1. Can the context and flag be combined into one class?
+ *   - i.e. Does it make sense to one a feature-flag apply to two different context types?
+ * 2. Should the default value be merged into the Flag type instead of specified on each check?
+ *   - Or maybe still allow it to be specified, but make it an optional parameter
+ */
+
+/**
+ * Feature-Flag Client interface.
+ */
+sealed interface FeatureFlagClient {
+    /**
+     * Returns true if the flag with the provided context should be enabled. Returns false otherwise.
+     */
+    fun enabled(flag: Flag, ctx: Context): Boolean
+}
+
+/**
+ * Platform feature-flag client. Feature-flag are derived from a yaml config file.
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [config] the location of the yaml config file that contains the feature-flag definitions.
+ * The [config] will be watched for changes and the internal representation of the [config] will be updated to match.
+ */
+class PlatformClient(config: Path) : FeatureFlagClient {
+    /** [flags] holds the mappings of the flag-name to the flag properties */
+    private var flags: Map<String, PlatformFlag> = readConfig(config)
+
+    /** lock is used for ensuring access to the flags map is handled correctly when the map is being updated. */
+    private val lock = ReentrantReadWriteLock()
+
+    init {
+        if (!config.isRegularFile()) {
+            throw IllegalArgumentException("config must reference a file")
+        }
+
+        config.onChange {
+            val tempFlags = readConfig(config)
+            lock.write { flags = tempFlags }
+        }
+    }
+
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
```

Comment:
Should we add another function to return a string or another value type. We currently have a string feature flag: https://github.com/airbytehq/airbyte/blob/master/airbyte-commons/src/main/java/io/airbyte/commons/features/FeatureFlags.java#L38 Maybe it should have been move the the EnvConfig though

Reply:
Good call, I missed this during my research.  I think it makes sense to add in a string version as well, given that we already appear to be using it.

Reply:
@benmoriceau I dug into this a little further and I don't think we need to support string feature-flags yet. The `FIELD_SELECTION_WORKSPACES` contains a comma-delimited list of workspace identifiers that should have field selection enabled.  This maps to the `Workspace` Context concept that this PR introduces.

Currently the env-var is defined similar to `FIELD_SELECTION_WORKSPACES: "uuid-0000, uuid-0001"`
which is used to determine which workspaces should have this functionality enabled.  Instead of having the workspace-ids defined in this configuration you would grab the current workspace-id, wrap it in the `Workspace` context and then pass it to the feature-flag client. 

The hard-coded workspace-ids would now reside on the LD side (and would be evaluated against the `Workspace` context provided) instead of being defined within the configuration.

Basically, in LD, if the provided workspace (from the context) matches one of the workspaces listed in LD, then that feature-flag would evaluate to `true`, otherwise `false`.